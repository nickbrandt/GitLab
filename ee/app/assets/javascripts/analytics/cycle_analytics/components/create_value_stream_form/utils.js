import { isEqual, pick } from 'lodash';
import { convertObjectPropsToSnakeCase } from '~/lib/utils/common_utils';
import { isStartEvent, getAllowedEndEvents, eventToOption, eventsByIdentifier } from '../../utils';
import {
  i18n,
  ERRORS,
  defaultErrors,
  defaultFields,
  NAME_MAX_LENGTH,
  formFieldKeys,
} from './constants';

/**
 * @typedef {Object} CustomStageEvents
 * @property {String} canBeStartEvent - Title of the metric measured
 * @property {String} name - Friendly name for the event
 * @property {String} identifier - snakeized name for the event
 *
 * @typedef {Object} DropdownData
 * @property {String} text - Friendly name for the event
 * @property {String} value - Value to be submitted for the dropdown
 */

/**
 * Takes an array of custom stage events to return only the
 * events where `canBeStartEvent` is true and converts them
 * to { value, text } pairs for use in dropdowns
 *
 * @param {CustomStageEvents[]} events
 * @returns {DropdownData[]} array of start events formatted for dropdowns
 */
export const startEventOptions = (eventsList) => [
  { value: null, text: i18n.SELECT_START_EVENT },
  ...eventsList.filter(isStartEvent).map(eventToOption),
];

/**
 * Takes an array of custom stage events to return only the
 * events where `canBeStartEvent` is false and converts them
 * to { value, text } pairs for use in dropdowns
 *
 * @param {CustomStageEvents[]} events
 * @returns {DropdownData[]} array end events formatted for dropdowns
 */
export const endEventOptions = (eventsList, startEventIdentifier) => {
  const endEvents = getAllowedEndEvents(eventsList, startEventIdentifier);
  return [
    { value: null, text: i18n.SELECT_END_EVENT },
    ...eventsByIdentifier(eventsList, endEvents).map(eventToOption),
  ];
};

/**
 * @typedef {Object} CustomStageFormData
 * @property {Object.<String, String>} fields - form field values
 * @property {Object.<String, Array>} errors - form field errors
 */

/**
 * Initializes the fields and errors for the custom stages form
 * providing defaults for any missing keys
 *
 * @param {CustomStageFormData} data
 * @returns {CustomStageFormData} the updated initial data with all defaults
 */
export const initializeFormData = ({ fields, errors }) => {
  const initErrors = fields?.endEventIdentifier
    ? defaultErrors
    : {
        ...defaultErrors,
        endEventIdentifier: !fields?.startEventIdentifier ? [ERRORS.START_EVENT_REQUIRED] : [],
      };
  return {
    fields: {
      ...defaultFields,
      ...fields,
    },
    errors: {
      ...initErrors,
      ...errors,
    },
  };
};

/**
 * Validates the form fields for the custom stages form
 * Any errors will be returned in a object where the key is
 * the name of the field.g
 *
 * @param {Object} fields key value pair of form field values
 * @param {Object} defaultStageNames array of lower case default value stream names
 * @returns {Object} key value pair of form fields with an array of errors
 */
export const validateStage = (fields, defaultStageNames = []) => {
  const newErrors = {};

  if (fields?.name) {
    if (fields.name.length > NAME_MAX_LENGTH) {
      newErrors.name = [ERRORS.MAX_LENGTH];
    }
    if (fields?.custom && defaultStageNames.includes(fields.name.toLowerCase())) {
      newErrors.name = [ERRORS.STAGE_NAME_EXISTS];
    }
  } else {
    newErrors.name = [ERRORS.STAGE_NAME_MIN_LENGTH];
  }

  if (fields?.startEventIdentifier) {
    if (!fields?.endEventIdentifier) {
      newErrors.endEventIdentifier = [ERRORS.END_EVENT_REQUIRED];
    }
  } else {
    newErrors.endEventIdentifier = [ERRORS.START_EVENT_REQUIRED];
  }
  return newErrors;
};

/**
 * Validates the name of a value stream Any errors will be
 * returned as an array in a object with key`name`
 *
 * @param {Object} fields key value pair of form field values
 * @returns {Array} an array of errors
 */
export const validateValueStreamName = ({ name = '' }) => {
  const errors = [];
  if (name.length > NAME_MAX_LENGTH) {
    errors.push(ERRORS.MAX_LENGTH);
  }
  if (!name.length) {
    errors.push(ERRORS.VALUE_STREAM_NAME_MIN_LENGTH);
  }
  return errors;
};

/**
 * Formats the value stream stages for submission, ensures that the
 * 'custom' property is set when we are editing, we include the `id` if its
 * set and all fields are converted to snake case
 *
 * @param {Array} stages array of value stream stages
 * @param {Boolean} isEditing flag to indicate if we are editing a value stream or creating
 * @returns {Array} the array prepared to be submitted for persistence
 */
export const formatStageDataForSubmission = (stages, isEditing = false) => {
  return stages.map(({ id = null, custom = false, name, ...rest }) => {
    let editProps = { custom };
    if (isEditing) {
      // We can add a new stage to the value stream when either creating, or editing
      // If a new stage has been added then at this point, the `id` won't exist
      // The new stage is still `custom` but wont have an id until the form submits and its persisted to the DB
      editProps = id ? { id, custom: true } : { custom: true };
    }
    // While we work on https://gitlab.com/gitlab-org/gitlab/-/issues/321959 we should not allow editing default
    return custom
      ? convertObjectPropsToSnakeCase({ ...rest, ...editProps, name })
      : convertObjectPropsToSnakeCase({ ...editProps, name, custom: false });
  });
};

/**
 * Checks an array of value stream stages to see if there are
 * any differences in the values they contain
 *
 * @param {Array} stages array of value stream stages
 * @param {Array} stages array of value stream stages
 * @returns {Boolean} returns true if there is a difference in the 2 arrays
 */
export const hasDirtyStage = (currentStages, originalStages) => {
  const cs = currentStages.map((s) => pick(s, formFieldKeys));
  const os = originalStages.map((s) => pick(s, formFieldKeys));
  return !isEqual(cs, os);
};

/**
 * Checks if the target name matches the name of any of the value
 * stream stages passed in
 *
 * @param {Array} stages array of value stream stages
 * @param {String} targetName name we are searching for
 * @returns {Object} returns the found object or null
 */
const findStageByName = (stages, targetName = '') =>
  stages.find(({ name }) => name.toLowerCase().trim() === targetName.toLowerCase().trim());

/**
 * Returns a valid custom value stream stage
 *
 * @param {Object} stage a raw value stream stage retrieved from the vuex store
 * @returns {Object} the same stage with fields adjusted for the value stream form
 */
const prepareCustomStage = ({ startEventLabel = {}, endEventLabel = {}, ...rest }) => ({
  ...rest,
  startEventLabel,
  endEventLabel,
  startEventLabelId: startEventLabel?.id || null,
  endEventLabelId: endEventLabel?.id || null,
  isDefault: false,
});

/**
 * Returns a valid default value stream stage
 *
 * @param {Object} stage a raw value stream stage retrieved from the vuex store
 * @returns {Object} the same stage with fields adjusted for the value stream form
 */
const prepareDefaultStage = (defaultStageConfig, { name, ...rest }) => {
  // default stages currently dont have any label based events
  const stage = findStageByName(defaultStageConfig, name) || null;
  if (!stage) return {};
  const { startEventIdentifier = null, endEventIdentifier = null } = stage;
  return {
    ...rest,
    name,
    startEventIdentifier,
    endEventIdentifier,
    isDefault: true,
  };
};

const generateHiddenDefaultStages = (defaultStageConfig, stageNames) => {
  // We use the stage name to check for any default stages that might be hidden
  // Currently the default stages can't be renamed
  return defaultStageConfig
    .filter(({ name }) => !stageNames.includes(name.toLowerCase()))
    .map((data) => ({ ...data, hidden: true }));
};

/**
 * Returns a valid array of value stream stages for
 * use in the value stream form
 *
 * @param {Array} defaultStageConfig an array of the default value stream stages retrieved from the backend
 * @param {Array} selectedValueStreamStages an array of raw value stream stages retrieved from the vuex store
 * @returns {Object} the same stage with fields adjusted for the value stream form
 */
export const generateInitialStageData = (defaultStageConfig, selectedValueStreamStages) => {
  const hiddenDefaultStages = generateHiddenDefaultStages(
    defaultStageConfig,
    selectedValueStreamStages.map((s) => s.name.toLowerCase()),
  );
  const combinedStages = [...selectedValueStreamStages, ...hiddenDefaultStages];
  return combinedStages.map(
    ({ startEventIdentifier = null, endEventIdentifier = null, custom = false, ...rest }) => {
      const stageData =
        custom && startEventIdentifier && endEventIdentifier
          ? prepareCustomStage({ ...rest, startEventIdentifier, endEventIdentifier })
          : prepareDefaultStage(defaultStageConfig, rest);

      if (stageData?.name) {
        return {
          ...stageData,
          custom,
        };
      }
      return {};
    },
  );
};
