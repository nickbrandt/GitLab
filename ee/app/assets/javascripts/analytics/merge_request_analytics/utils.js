import dateFormat from 'dateformat';
import { dateFormats } from '~/analytics/shared/constants';
import {
  getMonthNames,
  getDateInPast,
  getDayDifference,
  secondsToDays,
} from '~/lib/utils/datetime_utility';
import { THROUGHPUT_CHART_STRINGS, DEFAULT_NUMBER_OF_DAYS, UNITS } from './constants';

/**
 * A utility function which accepts a date range and returns
 * computed month data which is required to build the GraphQL
 * query for the Throughput Analytics chart
 *
 * @param {Date} startDate the startDate for the data range
 * @param {Date} endDate the endDate for the data range
 * @param {String} format the date format to be used
 *
 * @return {Array} the computed month data
 */
export const computeMonthRangeData = (startDate, endDate, format = dateFormats.isoDate) => {
  const monthData = [];
  const monthNames = getMonthNames(true);

  for (
    let dateCursor = new Date(endDate);
    dateCursor >= startDate;
    dateCursor.setMonth(dateCursor.getMonth(), 0)
  ) {
    const monthIndex = dateCursor.getMonth();
    const year = dateCursor.getFullYear();

    const mergedAfter = new Date(year, monthIndex, 1);
    const mergedBefore = new Date(year, monthIndex + 1, 1);

    monthData.unshift({
      year,
      month: monthNames[monthIndex],
      mergedAfter: dateFormat(mergedAfter, format),
      mergedBefore: dateFormat(mergedBefore, format),
    });
  }

  if (monthData.length) {
    monthData[0].mergedAfter = dateFormat(startDate, format); // Set first item to startDate
    monthData[monthData.length - 1].mergedBefore = dateFormat(endDate, format); // Set last item to endDate
  }

  return monthData;
};

/**
 * A utility function which accepts the raw throughput chart data
 * and transforms it into the format required for the area chart.
 *
 * @param {Object} chartData the raw chart data
 *
 * @return {Array} the formatted chart data
 */
export const formatThroughputChartData = (chartData) => {
  if (!chartData) return [];

  const data = Object.keys(chartData)
    .slice(0, -1) // Remove the __typeName key
    .map((value) => [value.split('_').join(' '), chartData[value].count]); // key: Aug_2020 => Aug 2020

  return [
    {
      name: THROUGHPUT_CHART_STRINGS.Y_AXIS_TITLE,
      data,
    },
  ];
};

/**
 * A utility function which accepts the raw throughput data
 * and computes the mean time to merge.
 *
 * @param {Object} rawData the raw throughput data
 *
 * @return {Object} the computed MTTM data
 */
export const computeMttmData = (rawData) => {
  if (!rawData) return {};

  const mttmData = Object.values(rawData)
    // eslint-disable-next-line @gitlab/require-i18n-strings
    .filter((value) => value !== 'Project')
    .reduce(
      (total, monthData) => {
        return {
          count: total.count + monthData.count,
          totalTimeToMerge: total.totalTimeToMerge + monthData.totalTimeToMerge,
        };
      },
      {
        count: 0,
        totalTimeToMerge: 0,
      },
    );

  // GlSingleStat expects a String for the 'value' prop
  // https://gitlab.com/gitlab-org/gitlab-ui/-/issues/1152
  return {
    title: THROUGHPUT_CHART_STRINGS.MTTM,
    value: `${secondsToDays(mttmData.totalTimeToMerge / mttmData.count)}`,
    unit: UNITS.DAYS,
  };
};

/**
 * A utility function which accepts start and end date params
 * and validates that the date range does not exceed the bounds
 *
 * @param {Date} startDate the startDate for the data range
 * @param {Date} endDate the endDate for the data range
 *
 * @return {Object} an object containing the startDate and endDate
 */
export const parseAndValidateDates = (startDateParam, endDateParam) => {
  let startDate = new Date(startDateParam);
  let endDate = new Date(endDateParam);
  const numberOfDays = getDayDifference(startDate, endDate);

  if (!startDateParam.length || numberOfDays > DEFAULT_NUMBER_OF_DAYS || endDate < startDate) {
    startDate = getDateInPast(new Date(), DEFAULT_NUMBER_OF_DAYS);
    endDate = new Date();
  }

  return { startDate, endDate };
};
