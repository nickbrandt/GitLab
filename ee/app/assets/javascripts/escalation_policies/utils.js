/**
 * Returns `true` for non-empty string, otherwise returns `false`
 * @param {String} name
 *
 * @returns {Boolean}
 */
export const isNameFieldValid = (name) => {
  return Boolean(name?.length);
};

/**
 * Returns an array of booleans  - validation state for each rule
 * @param {Array} rules
 *
 * @returns {Array}
 */
export const getRulesValidationState = (rules) => {
  return rules.map((rule) => {
    return {
      isTimeValid: parseInt(rule.elapsedTimeMinutes, 10) >= 0,
      isScheduleValid: Boolean(rule.oncallScheduleIid),
    };
  });
};

/**
 * Serializes a rule by converting elapsed minutes to seconds
 * @param {Object} rule
 *
 * @returns {Object} rule
 */
export const serializeRule = ({ elapsedTimeMinutes, ...ruleParams }) => ({
  ...ruleParams,
  elapsedTimeSeconds: elapsedTimeMinutes * 60,
});

/**
 * Parses a policy by converting elapsed seconds to minutes
 * @param {Object} policy
 *
 * @returns {Object} policy
 */
export const parsePolicy = (policy) => ({
  ...policy,
  rules: policy.rules.map(({ elapsedTimeSeconds, ...ruleParams }) => ({
    ...ruleParams,
    elapsedTimeMinutes: elapsedTimeSeconds / 60,
  })),
});
