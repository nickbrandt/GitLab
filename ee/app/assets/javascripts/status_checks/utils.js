export const modalPrimaryActionProps = (text, loading) => ({
  text,
  attributes: [{ variant: 'confirm', loading }],
});
