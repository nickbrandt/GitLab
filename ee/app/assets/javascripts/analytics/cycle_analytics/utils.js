import dateFormat from 'dateformat';
import { isNumber } from 'lodash';
import { dateFormats } from '~/analytics/shared/constants';
import { OVERVIEW_STAGE_ID } from '~/cycle_analytics/constants';
import { medianTimeToParsedSeconds } from '~/cycle_analytics/utils';
import createFlash, { hideFlash } from '~/flash';
import { convertObjectPropsToCamelCase } from '~/lib/utils/common_utils';
import { newDate, dayAfter, secondsToDays, getDatesInRange } from '~/lib/utils/datetime_utility';
import httpStatus from '~/lib/utils/http_status';
import { convertToSnakeCase, slugify } from '~/lib/utils/text_utility';
import { toYmd } from '../shared/utils';

const EVENT_TYPE_LABEL = 'label';

export const removeFlash = (type = 'alert') => {
  const flashEl = document.querySelector(`.flash-${type}`);
  if (flashEl) {
    hideFlash(flashEl);
  }
};

export const toggleSelectedLabel = ({ selectedLabelIds = [], value = null }) => {
  if (!value) return selectedLabelIds;
  return selectedLabelIds.includes(value)
    ? selectedLabelIds.filter((v) => v !== value)
    : [...selectedLabelIds, value];
};

export const isStartEvent = (ev) =>
  Boolean(ev) && Boolean(ev.canBeStartEvent) && ev.canBeStartEvent;

export const eventToOption = (obj = null) => {
  if (!obj || (!obj.text && !obj.identifier)) return null;
  const { name: text = '', identifier: value = null } = obj;
  return { text, value };
};

export const getAllowedEndEvents = (events = [], targetIdentifier = null) => {
  if (!targetIdentifier || !events.length) return [];
  const st = events.find(({ identifier }) => identifier === targetIdentifier);
  return st && st.allowedEndEvents ? st.allowedEndEvents : [];
};

export const eventsByIdentifier = (events = [], targetIdentifier = []) => {
  if (!targetIdentifier || !targetIdentifier.length || !events.length) return [];
  return events.filter(({ identifier = '' }) => targetIdentifier.includes(identifier));
};

export const isLabelEvent = (labelEvents = [], ev = null) =>
  Boolean(ev) && labelEvents.length && labelEvents.includes(ev);

export const getLabelEventsIdentifiers = (events = []) =>
  events.filter((ev) => ev.type && ev.type === EVENT_TYPE_LABEL).map((i) => i.identifier);

/**
 * Checks if the specified stage is in memory or persisted to storage based on the id
 *
 * Default value stream analytics stages are initially stored in memory, when they are first
 * created the id for the stage is the name of the stage in lowercase. This string id
 * is used to fetch stage data (events, median calculation)
 *
 * When either a custom stage is created or an edit is made to a default stage then the
 * default stages get persisted to storage and will have a numeric id. The new numeric
 * id should then be used to access stage data
 *
 */
export const isPersistedStage = ({ custom, id }) => custom || isNumber(id);

/**
 * Returns the the correct slug to use for a stage
 * default stages use the snakecased title of the stage, while custom
 * stages will have a numeric id
 *
 * @param {Object} obj
 * @param {string} obj.title - title of the stage
 * @param {number} obj.id - numerical object id available for custom stages
 * @param {boolean} obj.custom - boolean flag indicating a custom stage
 * @returns {(number|string)} Returns a numerical id for customs stages and string for default stages
 */
const stageUrlSlug = ({ id, title, custom = false }) => {
  if (custom) return id;

  return convertToSnakeCase(title);
};

export const transformRawStages = (stages = []) =>
  stages.map(({ id, title, name = '', custom = false, ...rest }) => ({
    ...convertObjectPropsToCamelCase(rest, { deep: true }),
    id,
    title,
    custom,
    slug: isPersistedStage({ custom, id }) ? id : stageUrlSlug({ custom, id, title }),
    // the name field is used to create a stage, but the get request returns title
    name: name.length ? name : title,
  }));

export const transformRawTasksByTypeData = (data = []) => {
  if (!data.length) return [];
  return data.map((d) => convertObjectPropsToCamelCase(d, { deep: true }));
};

/**
 * Prepares the stage errors for use in the create value stream form
 *
 * The JSON error response returns a key value pair, the key corresponds to the
 * index of the stage with errors and the value is the returned error(s)
 *
 * @param {Array} stages - Array of value stream stages
 * @param {Object} errors - Key value pair of stage errors
 * @returns {Array} Returns and array of stage error objects
 */
export const prepareStageErrors = (stages, errors) =>
  stages.length ? stages.map((_, index) => convertObjectPropsToCamelCase(errors[index]) || {}) : [];

/**
 * Takes the duration data for selected stages, transforms the date values and returns
 * the data in a flattened array
 *
 * The received data is expected to be the following format; One top level object in the array per stage,
 * each potentially having multiple data entries.
 * [
 *   {
 *    slug: 'issue',
 *    selected: true,
 *    data: [
 *      {
 *        'average_duration_in_seconds': 1234,
 *        'date': '2019-09-02T18:25:43.511Z'
 *      },
 *      ...
 *    ]
 *   },
 *   ...
 * ]
 *
 * The data is then transformed and flattened into the following format;
 * [
 *  {
 *    'average_duration_in_seconds': 1234,
 *    'date': '2019-09-02'
 *  },
 *  ...
 * ]
 *
 * @param {Array} data - The duration data for selected stages
 * @returns {Array} An array with each item being an object containing the average_duration_in_seconds and date values for an event
 */
export const flattenDurationChartData = (data) =>
  data
    .map((stage) =>
      stage.data.map((event) => {
        const date = new Date(event.date);
        return {
          ...event,
          date: dateFormat(date, dateFormats.isoDate),
        };
      }),
    )
    .flat();

/**
 * Takes the duration data for selected stages, groups the data by day and calculates the average duration
 * per day, for stages with values on that specific day.
 *
 * The received data is expected to be the following format; One top level object in the array per stage,
 * each potentially having multiple data entries.
 * [
 *   {
 *    slug: 'issue',
 *    selected: true,
 *    data: [
 *      {
 *        'average_duration_in_seconds': 1234,
 *        'date': '2019-09-02T18:25:43.511Z'
 *      },
 *      ...
 *    ]
 *   },
 *   ...
 * ]
 *
 * The data is then computed and transformed into a format that can be passed to the chart:
 * [
 *  ['2019-09-02', 7, '2019-09-02'],
 *  ['2019-09-03', 10, '2019-09-03'],
 *  ['2019-09-04', 8, '2019-09-04'],
 *  ...
 * ]
 *
 * In the data above, each array i represents a point in the scatterplot with the following data:
 * i[0] = date, displayed on x axis
 * i[1] = metric, displayed on y axis
 * i[2] = date, used in the tooltip
 *
 * @param {Array} data - The duration data for selected stages
 * @param {Date} startDate - The globally selected Value Stream Analytics start date
 * @param {Date} endDate - The globally selected Value Stream Analytics end date
 * @returns {Array} An array with each item being another arry of three items (plottable date, computed average, tooltip display date)
 */
export const getDurationChartData = (data, startDate, endDate) => {
  const flattenedData = flattenDurationChartData(data);
  const eventData = [];
  const endOfDay = newDate(endDate);
  endOfDay.setHours(23, 59, 59); // make sure we're at the end of the day

  for (
    let currentDate = newDate(startDate);
    currentDate <= endOfDay;
    currentDate = dayAfter(currentDate)
  ) {
    const currentISODate = dateFormat(newDate(currentDate), dateFormats.isoDate);
    const valuesForDay = flattenedData.filter((object) => object.date === currentISODate);
    const averagedData =
      valuesForDay.reduce((total, value) => total + value.average_duration_in_seconds, 0) /
      valuesForDay.length;
    const averagedDataInDays = secondsToDays(averagedData);

    if (averagedDataInDays) eventData.push([currentISODate, averagedDataInDays, currentISODate]);
  }

  return eventData;
};

export const orderByDate = (a, b, dateFmt = (datetime) => new Date(datetime).getTime()) =>
  dateFmt(a) - dateFmt(b);

/**
 * Takes a dictionary of dates and the associated value, sorts them and returns just the value
 *
 * @param {Object.<Date, number>} series - Key value pair of dates and the value for that date
 * @returns {number[]} The values of each key value pair
 */
export const flattenTaskByTypeSeries = (series = {}) =>
  Object.entries(series)
    .sort((a, b) => orderByDate(a[0], b[0]))
    .map((dataSet) => dataSet[1]);

/**
 * @typedef {Object} RawTasksByTypeData
 * @property {Object} label - Raw data for a group label
 * @property {Array} series - Array of arrays with date and associated value ie [ ['2020-01-01', 10],['2020-01-02', 10] ]

 * @typedef {Object} TransformedTasksByTypeData
 * @property {Array} groupBy - The list of dates for the range of data in each data series
 * @property {Array} data - An array of the data values for each series
 * @property {Array} seriesNames - Names of the series to be charted ie label names
 */

/**
 * Takes the raw tasks by type data and generates an array of data points,
 * an array of data series and an array of data labels for the given time period.
 *
 * Currently the data is transformed to support use in a stacked column chart:
 * https://gitlab-org.gitlab.io/gitlab-ui/?path=/story/charts-stacked-column-chart--stacked
 *
 * @param {Object} obj
 * @param {RawTasksByTypeData[]} obj.data - array of raw data, each element contains a label and series
 * @param {Date} obj.startDate - start date in ISO date format
 * @param {Date} obj.endDate - end date in ISO date format
 *
 * @returns {TransformedTasksByTypeData} The transformed data ready for use in charts
 */
export const getTasksByTypeData = ({ data = [], startDate = null, endDate = null }) => {
  if (!startDate || !endDate || !data.length) {
    return {
      groupBy: [],
      data: [],
    };
  }

  const groupBy = getDatesInRange(startDate, endDate, toYmd).sort(orderByDate);
  const zeroValuesForEachDataPoint = groupBy.reduce(
    (acc, date) => ({
      ...acc,
      [date]: 0,
    }),
    {},
  );

  const transformed = data.reduce(
    (acc, curr) => {
      const {
        label: { title: name },
        series,
      } = curr;
      acc.data = [
        ...acc.data,
        {
          name,
          // adds 0 values for each data point and overrides with data from the series
          data: flattenTaskByTypeSeries({
            ...zeroValuesForEachDataPoint,
            ...Object.fromEntries(series),
          }),
        },
      ];
      return acc;
    },
    {
      data: [],
      seriesNames: [],
    },
  );

  return {
    ...transformed,
    groupBy,
  };
};

const buildDataError = ({ status = httpStatus.INTERNAL_SERVER_ERROR, error }) => {
  const err = new Error(error);
  err.errorCode = status;
  return err;
};

/**
 * Flashes an error message if the status code is not 200
 *
 * @param {Object} error - Axios error object
 * @param {String} errorMessage - Error message to display
 */
export const flashErrorIfStatusNotOk = ({ error, message }) => {
  if (error?.errorCode !== httpStatus.OK) {
    createFlash({
      message,
    });
  }
};

/**
 * Data errors can occur when DB queries for analytics data time out
 * The server will respond with a status `200` success and include the
 * relevant error in the response body
 *
 * @param {Object} Response - Axios ajax response
 * @returns {Object} Returns the axios ajax response
 */
export const checkForDataError = (response) => {
  const { data, status } = response;
  if (data?.error) {
    throw buildDataError({ status, error: data.error });
  }
  return response;
};

export const throwIfUserForbidden = (error) => {
  if (error?.response?.status === httpStatus.FORBIDDEN) {
    throw error;
  }
};

/**
 * Takes the raw median value arrays and converts them into a useful object
 * containing the string for display in the path navigation, additionally
 * the overview is calculated as a sum of all the stages.
 * ie. converts [{ id: 'test', value: 172800 }] => { 'test': '2d' }
 *
 * @param {Array} Medians - Array of stage median objects, each contains a `id`, `value` and `error`
 * @returns {Object} Returns key value pair with the stage name and its display median value
 */
export const formatMedianValuesWithOverview = (medians = []) => {
  const calculatedMedians = medians.reduce(
    (acc, { id, value = 0 }) => {
      return {
        ...acc,
        [id]: value ? medianTimeToParsedSeconds(value) : '-',
        [OVERVIEW_STAGE_ID]: acc[OVERVIEW_STAGE_ID] + value,
      };
    },
    {
      [OVERVIEW_STAGE_ID]: 0,
    },
  );
  const overviewMedian = calculatedMedians[OVERVIEW_STAGE_ID];
  return {
    ...calculatedMedians,
    [OVERVIEW_STAGE_ID]: overviewMedian ? medianTimeToParsedSeconds(overviewMedian) : '-',
  };
};

/**
 * @typedef {Object} MetricData
 * @property {String} title - Title of the metric measured
 * @property {String} value - String representing the decimal point value, e.g '1.5'
 * @property {String} [unit] - String representing the decimal point value, e.g '1.5'
 *
 * @typedef {Object} TransformedMetricData
 * @property {String} label - Title of the metric measured
 * @property {String} value - String representing the decimal point value, e.g '1.5'
 * @property {String} key - Slugified string based on the 'title'
 * @property {String} description - String to display for a description
 * @property {String} unit - String representing the decimal point value, e.g '1.5'
 */

/**
 * Prepares metric data to be rendered in the metric_card component
 *
 * @param {MetricData[]} data - The metric data to be rendered
 * @param {Object} popoverContent - Key value pair of data to display in the popover
 * @returns {TransformedMetricData[]} An array of metrics ready to render in the metric_card
 */

export const prepareTimeMetricsData = (data = [], popoverContent = {}) =>
  data.map(({ title: label, ...rest }) => {
    const key = slugify(label);
    return {
      ...rest,
      label,
      key,
      description: popoverContent[key]?.description || '',
    };
  });
