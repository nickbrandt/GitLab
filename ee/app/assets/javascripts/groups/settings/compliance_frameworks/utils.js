import Api from '~/api';
import httpStatus from '~/lib/utils/http_status';
import { isNumeric } from '~/lib/utils/number_utils';
import { EDIT_PATH_ID_FORMAT, PIPELINE_CONFIGURATION_PATH_FORMAT } from './constants';

export const injectIdIntoEditPath = (path, id) => {
  if (!path || !path.match(EDIT_PATH_ID_FORMAT) || !isNumeric(id)) {
    return '';
  }

  return path.replace(EDIT_PATH_ID_FORMAT, `/${id}/`);
};

export const initialiseFormData = () => ({
  name: null,
  description: null,
  pipelineConfigurationFullPath: null,
  color: null,
});

export const getSubmissionParams = (formData, pipelineConfigurationFullPathEnabled) => {
  const params = { ...formData };

  if (!pipelineConfigurationFullPathEnabled) {
    delete params.pipelineConfigurationFullPath;
  }

  return params;
};

export const getPipelineConfigurationPathParts = (path) => {
  const [, file, group, project] = path.match(PIPELINE_CONFIGURATION_PATH_FORMAT) || [];

  return { file, group, project };
};

export const validatePipelineConfirmationFormat = (path) =>
  PIPELINE_CONFIGURATION_PATH_FORMAT.test(path);

export const fetchPipelineConfigurationFileExists = async (path) => {
  const { file, group, project } = getPipelineConfigurationPathParts(path);

  if (!file || !group || !project) {
    return false;
  }

  try {
    const { status } = await Api.getRawFile(`${group}/${project}`, file);

    return status === httpStatus.OK;
  } catch (e) {
    return false;
  }
};
