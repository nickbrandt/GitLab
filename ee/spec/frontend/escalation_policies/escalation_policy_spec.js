import { GlSprintf } from '@gitlab/ui';
import { shallowMount } from '@vue/test-utils';
import { cloneDeep } from 'lodash';
import EditEscalationPolicyModal from 'ee/escalation_policies/components/add_edit_escalation_policy_modal.vue';
import DeleteEscalationPolicyModal from 'ee/escalation_policies/components/delete_escalation_policy_modal.vue';
import EscalationPolicy from 'ee/escalation_policies/components/escalation_policy.vue';

import {
  deleteEscalationPolicyModalId,
  editEscalationPolicyModalId,
} from 'ee/escalation_policies/constants';
import { parsePolicy } from 'ee/escalation_policies/utils';
import mockPolicies from './mocks/mockPolicies.json';

describe('EscalationPolicy', () => {
  let wrapper;
  const escalationPolicy = parsePolicy(cloneDeep(mockPolicies[0]));

  const createComponent = () => {
    wrapper = shallowMount(EscalationPolicy, {
      propsData: {
        policy: escalationPolicy,
        index: 0,
      },
      stubs: {
        GlSprintf,
      },
    });
  };

  beforeEach(() => {
    createComponent();
  });

  afterEach(() => {
    wrapper.destroy();
  });

  const findDeleteModal = () => wrapper.findComponent(DeleteEscalationPolicyModal);
  const findEditModal = () => wrapper.findComponent(EditEscalationPolicyModal);

  it('renders policy with rules', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('Modals', () => {
    describe('delete policy modal', () => {
      it('should render a modal and provide it with correct id', () => {
        const modal = findDeleteModal();
        expect(modal.exists()).toBe(true);
        expect(modal.props('modalId')).toBe(
          `${deleteEscalationPolicyModalId}-${escalationPolicy.id}`,
        );
      });
    });

    describe('edit policy modal', () => {
      it('should render a modal and provide it with correct id and isEditMode props', () => {
        const modal = findEditModal();
        expect(modal.exists()).toBe(true);
        expect(modal.props('modalId')).toBe(
          `${editEscalationPolicyModalId}-${escalationPolicy.id}`,
        );
        expect(modal.props('isEditMode')).toBe(true);
      });
    });
  });
});
