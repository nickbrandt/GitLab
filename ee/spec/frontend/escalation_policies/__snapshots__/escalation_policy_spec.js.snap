// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EscalationPolicy renders policy with rules 1`] = `
<div>
  <gl-card-stub
    bodyclass="[object Object]"
    class="gl-mt-5"
    footerclass=""
    headerclass="[object Object]"
  >
     
    <gl-collapse-stub
      visible="true"
    >
      <p
        class="gl-text-gray-500 gl-mb-5"
      >
        
        Description 1 lives here
      
      </p>
       
      <div
        class="gl-border-solid gl-border-1 gl-border-gray-100 gl-rounded-base gl-p-5"
      >
        <div
          class="gl-mb-5"
        >
          <gl-icon-stub
            class="gl-mr-3"
            name="clock"
            size="16"
          />
           IF alert is not 
              acknowledged
             in 
          <span
            class="gl-font-weight-bold"
          >
            
                1 mins
              
          </span>
           
          <span
            class="right-arrow"
          >
            <i
              class="right-arrow-head"
            />
          </span>
           
          <gl-icon-stub
            class="gl-mr-3"
            name="notifications"
            size="16"
          />
           THEN 
              email on-call user in schedule
             
          <span
            class="gl-font-weight-bold"
          >
            
                Schedule to fill in
              
          </span>
        </div>
        <div
          class=""
        >
          <gl-icon-stub
            class="gl-mr-3"
            name="clock"
            size="16"
          />
           IF alert is not 
              resolved
             in 
          <span
            class="gl-font-weight-bold"
          >
            
                2 mins
              
          </span>
           
          <span
            class="right-arrow"
          >
            <i
              class="right-arrow-head"
            />
          </span>
           
          <gl-icon-stub
            class="gl-mr-3"
            name="notifications"
            size="16"
          />
           THEN 
              email on-call user in schedule
             
          <span
            class="gl-font-weight-bold"
          >
            
                Monitor schedule
              
          </span>
        </div>
      </div>
    </gl-collapse-stub>
  </gl-card-stub>
   
  <delete-escalation-policy-modal-stub
    escalationpolicy="[object Object]"
    modalid="deleteEscalationPolicyModal-37"
  />
   
  <edit-escalation-policy-modal-stub
    escalationpolicy="[object Object]"
    iseditmode="true"
    modalid="editEscalationPolicyModal-37"
  />
</div>
`;
